ergo-nipopow/src/nipopow_algos.rs

```diff
@@ -0,0 +1,351 @@
+///    `N == 0`.
+///
+/// Satoshi's original implementation used BN_bn2mpi() and BN_mpi2bn(). MPI uses the most
+/// significant bit of the first byte as sign. Thus 0x1234560000 is compact 0x05123456 and
+/// 0xc0de000000 is compact 0x0600c0de. Compact 0x05c0de00 would be -0x40de000000.
+///
+/// Bitcoin only uses this "compact" format for encoding difficulty targets, which are unsigned
+/// 256bit quantities.  Thus, all the complexities of the sign bit and using base 256 are probably
+/// an implementation accident.
+pub(crate) fn decode_compact_bits(n_bits: u64) -> BigInt {
+    let compact = n_bits as i64;
+    let size = ((compact >> 24) as i32) & 0xFF;
+    if size == 0 {
+        return BigInt::from(0);
+    }
+    let mut buf: Vec<i8> = std::iter::repeat(0).take(size as usize).collect();
+    if size >= 1 {
+        // Store the first byte of the mantissa
+        buf[0] = (((compact >> 16) as i32) & 0xFF) as i8;
+    }
+    if size >= 2 {
+        buf[1] = (((compact >> 8) as i32) & 0xFF) as i8;
+    }
+    if size >= 3 {
+        buf[2] = ((compact as i32) & 0xFF) as i8;
+    }
+
+    let is_negative = (buf[0] as i32) & 0x80 == 0x80;
+    if is_negative {
+        buf[0] &= 0x7f;
+        let buf: Vec<_> = buf.into_iter().map(|x| x as u8).collect();
+        -BigInt::from_signed_bytes_be(&buf)
+    } else {
+        let buf: Vec<_> = buf.into_iter().map(|x| x as u8).collect();
+        BigInt::from_signed_bytes_be(&buf)
+    }
+}
+
+/// This module exposes some internal code for integration testing.
+#[doc(hidden)]
+pub mod test {
+    use num_bigint::BigInt;
+
+    use crate::autolykos_pow_scheme::AutolykosPowScheme;
+
+    use super::decode_compact_bits;
+
+    /// Wrapper for `decode_compact_bits`
+    pub fn decode_n_bits(n_bits: u64) -> BigInt {
+        decode_compact_bits(n_bits)
+    }
+
+    /// Wrapper for `AutolykosPowScheme::calc_big_n` method
+    pub fn calc_big_n(header_version: u8, header_height: u32) -> usize {
+        let scheme = AutolykosPowScheme::default();
+        scheme.calc_big_n(header_version, header_height)
+    }
+
+    /// Constant data to be added to hash function to increase its calculation time
+    pub fn calc_big_m() -> Vec<u8> {
+        let scheme = AutolykosPowScheme::default();
+        scheme.calc_big_m()
+    }
+
+    /// Wrapper for `AutolykosPowScheme::calc_seed_v2` method
+    pub fn calc_seed_v2(
+        msg: &[u8],
+        big_n: usize,
+        nonce: &[u8],
+        header_height_bytes: &[u8],
+    ) -> Box<[u8; 32]> {
+        let scheme = AutolykosPowScheme::default();
+        #[allow(clippy::unwrap_used)]
+        scheme
+            .calc_seed_v2(big_n, msg, nonce, header_height_bytes)
+            .unwrap()
+    }
+
+    pub fn gen_indexes(seed_hash: &[u8; 32], big_n: usize) -> Vec<u32> {
+        let scheme = AutolykosPowScheme::default();
+        scheme.gen_indexes(seed_hash, big_n)
+    }
+}
```

Comment:
Created this module to expose some internal code for chain-generation in the `ergo-chain-generation` crate.

Reply:
I wonder if we should just make these internal methods into a public API of `nipopow_algos`. We're effectively exposing them as public here under new/same names anyway.

Reply:
Yes, I've made them public.